File: bongf/python/level2/kakao/2019/candidate_key.py
from itertools import combinations

## 일단 tuple의 원소들이 들어오는 순서가 같다고 가정 
## solution2보다 빠르다 
def solution(relation):
    x = len(relation)
    n = len(relation[0])
    candidates = []
    # print(n)
    for i in range(1, n+1):
        for c in combinations(range(0, n), i):
            candidates.append(list(c))
    
    cnt = 0
    print(candidates)
    for now, idxs in enumerate(candidates) :
        if(idxs == []):
            continue
        idxs = sorted(idxs)
        s1 = set()
        for r in relation:
            s2 = []
            for i in idxs:
                s2.append(r[i])
            s2 = tuple(s2)
            if s2 in s1 :
                break
            s1.add(s2)
        

        if(len(s1) < x):
            continue
        cnt += 1
        for a in range(now+1, len(candidates)):
            contain = True
            for b in idxs:
                if(b not in candidates[a]):
                    contain = False
                    break
            if contain :
                candidates[a] = []
     
    return cnt

## 4 = len(relation[0])
## 6 = len(relation)
# relation = [["100","ryan","music","2"],["200","apeach","math","2"],["300","tube","computer","3"],["400","con","computer","4"],["500","muzi","music","3"],["600","apeach","music","2"]]
# i = 1, 10, 11, 100, 101, 110, 111, 1000, 1001, ... 1100, 1101, 1110, 1111
# for i in range(1, 1<<4):
#     # print(f'i는 {i}')    
#     # print(bin(i))
#     tmp_set = set()
#     print(f'i는 {i}')
#     for j in range(6):
#         tmp = ''
#         for k in range(4):
#             # print(f'k는 {bin(1 << k)}')
#             ## i = 1001 이다 하면 relation[j]의 원소 0번째랑 네번째를 더해야 하잖아. 
#             ## 그래서 k 를 range(4로 돌면서) 비트연산으로 했을 때 둘이 겹치는 부분이 있는가 
#             # i = 1001 이면 k 비트연산 했을 때 k=0이면 1이니까 이 원소 포함해야 되서 더해야 하잖아.
#             # 그리고 k=1 이면 10이니까 해당 수 더하기 
#             # 그러니까 k의 역할을 i의 각자리수를 확인하면서 거기가 1인 친구를 relation 요소의 인덱스를 tmp에 더하는 역할 
#             ## 1 << k = 0001, 0010, 0100, 1000
#             if i & (1 << k):
#                 print("==맞대==")
#                 print(bin(i), bin(1<<k), str(relation[j][k]))
#                 tmp += str(relation[j][k])
         
#         # print(tmp)
#         tmp_set.add(tmp)
#         # print(tmp_set)
#     # print(tmp_set)

def solution2(relation):
    answer_list = list()
    for i in range(1, 1 << len(relation[0])):
        tmp_set = set()
        for j in range(len(relation)):
            tmp = ''
            for k in range(len(relation[0])):
                if i & (1 << k):
                    tmp += str(relation[j][k])
            tmp_set.add(tmp)

        if len(tmp_set) == len(relation):
            not_duplicate = True
            for num in answer_list:
                if (num & i) == num:
                    print(bin(num))
                    print(bin(i))
                    print(bin(num & i))
                    not_duplicate = False
                    break
            if not_duplicate:
                answer_list.append(i)
    return len(answer_list)



# print(solution([["100","ryan","music","2"],["200","apeach","math","2"],["300","tube","computer","3"],["400","con","computer","4"],["500","muzi","music","3"],["600","apeach","music","2"]]))
print(solution2([['a', 'aa'], ['aa', 'a'], ['a', 'a']])) ## 0, 이거 근데 왜 통과? 
print(solution([['a', 'aa'], ['aa', 'a'], ['a', 'a']])) ## 1 

File: programmers/java/src/level2/kakao/kakao2019/CandidateKey.java
package level2.kakao.kakao2019;

import java.util.*;

// https://programmers.co.kr/learn/courses/30/lessons/42890
public class CandidateKey {
    public List<List<Integer>> coms = new ArrayList<>();
    public int relationLength;
    public int dataLength;

    public static void main(String[] args) {
        CandidateKey ck = new CandidateKey();
        String s = "[['a', 'aa'], ['aa', 'a'], ['a', 'a']]";
//        System.out.println(s.replace("[", "{").replace("]", "}"));
        System.out.println(ck.solution3(new String[][]{{"100", "ryan", "music", "2"}, {"200", "apeach", "math", "2"}, {"300", "tube", "computer", "3"}, {"400", "con", "computer", "4"}, {"500", "muzi", "music", "3"}, {"600", "apeach", "music", "2"}}));
//        System.out.println(ck.solution3(new String[][]{{"a", "aa"}, {"aa", "a"}, {"a", "a"}}));
    }

    public int solution3(String[][] relations) {
        relationLength = relations.length;
        dataLength = relations[0].length;
        List<Integer> cands = new ArrayList<>();
        for (int bit = 0; bit < (1 << (dataLength + 1)); bit++) {
            Set<List<String>> midData = new HashSet<>();
            for (String[] rel : relations) {
                List<String> d = new ArrayList<>();
                for (int j = 0; j < dataLength; j++) {
                    if ((bit & (1 << j)) > 0) {
                        d.add(rel[j]);
                    }
                }
                midData.add(d);
            }
            if (midData.size() == relationLength) {
                cands.add(bit);
            }
        }

        boolean[] result = new boolean[cands.size()];
        Arrays.fill(result, true);
        for (int i = 0; i < result.length - 1; i++) {
            if (result[i]) {
                for (int j = i + 1; j < result.length; j++) {
                    if ((cands.get(i) & cands.get(j)) == cands.get(i)) {
                        result[j] = false;
                    }
                }
            }
        }

        int cnt = 0;
        for (boolean r : result) {
            if (r) {
                cnt++;
            }
        }
        return cnt;

    }


    public int solution(String[][] relation) {
        relationLength = relation.length;
        dataLength = relation[0].length;
        for (int i = 1; i <= dataLength; i++) {
            combination(i, 0, new Stack<Integer>());
        }
        System.out.println(coms);
        int answer = 0;
        for (int i = 0; i < coms.size(); i++) {
            List<Integer> idxs = coms.get(i);
            System.out.println(idxs);
            if (idxs.size() == 0) {
                continue;
            }
            Set<List<String>> set1 = new HashSet<>();
            for (String[] r : relation) {
                List<String> s2 = new ArrayList<>();
                for (Integer idx : idxs) {
                    s2.add(r[idx]);
                }
                if (set1.contains(s2)) {
                    break;
                }
                set1.add(s2);
            }
            if (set1.size() < relationLength) {
                continue;
            }
            answer++;
            for (int j = i + 1; j < coms.size(); j++) {
                boolean containAll = true;
                for (Integer idx : idxs) {
                    if (!coms.get(j).contains(idx)) {
                        containAll = false;
                        break;
                    }
                }
                if (containAll) {
                    coms.set(j, new ArrayList<>());
                }
            }
        }
        return answer;
    }

    public void combination(int toPick, int start, Stack<Integer> output) {
        if (toPick == 0) {
            coms.add(new ArrayList<>(output));
            return;
        }

        for (int i = start; i < dataLength; i++) {
            output.push(i);
            combination(toPick - 1, i + 1, output);
            output.pop();
        }
    }

    public int solution2(String[][] relation) {
        List<Integer> answers = new ArrayList<>();
        for (int i = 1; i < (1 << relation[0].length); i++) {
            Set<List<String>> set1 = new HashSet<>();
            for (int j = 0; j < relation.length; j++) {
                List<String> s2 = new ArrayList<>();
                for (int k = 0; k < relation[0].length; k++) {
                    if ((i & (1 << k)) != 0) {
                        s2.add(relation[j][k]);
                    }
                }
                set1.add(s2);
            }

            if (set1.size() == relation.length) {
                boolean toRemove = false;
                for (Integer answer : answers) {
                    if ((answer & i) == answer) {
                        toRemove = true;
                        break;
                    }
                }
                if (!toRemove) {
                    answers.add(i);
                }
            }
        }
        return answers.size();
    }
}
